### ISSUE TITLE
5.0.0.beta4 / actionpack - undefined method `symbol' for nil:NilClass*

### ISSUE THREAD

**Issue by philippfrank** [rails/rails#24827]
I cannot fully reproduce this issue but I am seeing some errors with the following details after upgrading to Rails 5:

**Error**
```
NoMethodError: undefined method `symbol' for nil:NilClass
```

**Location**
```
vendor/bundle/ruby/2.3.0/gems/actionpack-5.0.0.beta4/lib/action_dispatch/http/parameters.rb:63 - parse_formatted_parameters
```

**Request details** (from bugsnag)

*Headers*

```
{
  "Version": "HTTP/1.1",
  "Host": "...",
  "Connection": "close",
  "User-Agent": "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:46.0) Gecko/20100101 Firefox/46.0",
  "Cookie": "[FILTERED]",
  "X-Request-Id": "b883ebdb-95ac-4c81-89de-4885050fbd89",
  "X-Forwarded-For": "...",
  "X-Forwarded-Proto": "https",
  "X-Forwarded-Port": "443",
  "Via": "1.1 vegur",
  "Connect-Time": "0",
  "X-Request-Start": "1462260086212",
  "Total-Route-Time": "0",
  "Content-Length": "45"
}
```	

*Payload*

```
{
  "store_id": "a0d443bf-36ad-451d-8d9f-f7a5ef9b71fa"
}
```

Sorry, but I cannot provide you with more details.

**Comment by prathamesh-sonpatki** [comment]
@philippfrank This is happening because content type is not set on the request. Do you know why it is not set?

**Comment by philippfrank** [comment]
@prathamesh-sonpatki, thanks! I don't know why not, probably a browser thing. This code runs in production. I never saw a problem on Rails 4.

**Comment by prathamesh-sonpatki** [comment]
Investigating more why it is not working. A lot of things have changed related to params parsers in Rails 5.

**Comment by philippfrank** [comment]
@prathamesh-sonpatki Thanks

**Comment by vipulnsward** [comment]
@philippfrank did you see this on Rails5 beta3 or before ? Or you directly upgrade to beta4.

**Comment by philippfrank** [comment]
@vipulnsward I saw it both on Rails5 beta3 to which I upgraded from Rails4 and on Rails5 beta4.

**Comment by dopykuh** [comment]
Anybody a solution for this? It also appears in released version.

**Comment by rafaelfranca** [comment]
Can you please provide a sample application that reproduces the error?

**Comment by state-phasive** [comment]
I'm having the same issue.  It only reproduces if I do the following:

1. Generate an API project.
2. In controllers, add folder named 'v1'
3. Create controller 'user_controller.rb' in '/app/controllers/v1'
4. Add the following code:
`module V1
   class UserController < ApplicationController
      def do_something
         render json: { success: false }, status: :ok
      end
   end
end`
5. Add the following route to routes.rb:
`constraints subdomain: 'domainname' do
   scope 'v1' do
      post 'user/do_something', to: 'v1/user#do_something'
   end
end`
6. sudo nano /etc/hosts add:
`127.0.0.1        domainname.localhost.dev      domainname`

I am also using the following server gem:
Puma 3.4.0 in single mode.

I start rails like: `rails s -p 3400`

**Comment by state-phasive** [comment]
Once I added a header of:  `Content-type': 'application/json'`
It was able to get past the error.

**Comment by goya** [comment]
i am able to replicate this on the release version (5.0.0):

```
rails create blog
cd blog
bundle exec rails s
curl -H 'Content-Type:' -d 'data' http://localhost:3000
```

is the 500 the expected behavior for this? With rails 4.2.6 it gets passed to the router without throwing a 500 this low in the stack and can be handled by the user without (ugh) middleware.

**Comment by maclover7** [comment]
Bisecting and investigating.

**Comment by maclover7** [comment]
Bisected, the error is introduced (`NoMethodError: undefined method `symbol' for nil:NilClass`) via 97ed810cfc15725a0856227fa9f9eb26930f16c8. [Here](https://github.com/rails/rails/commit/97ed810cfc15725a0856227fa9f9eb26930f16c8#diff-6a80fa5836403509e86a81c2117af391R54), we are trying to get the symbol of a `nil` content type.

cc @meinac

**Comment by meinac** [comment]
Sorry for introducing this issue while solving another one. I couldn't realise the nil case.

cc @rafaelfranca

**Comment by maclover7** [comment]
@meinac no problem, happens to the best of us. would you be willing to work on a PR to fix the issue, and add a regression test?

**Comment by meinac** [comment]
At first glance I see that fixing this issue requires lots of changes. First of all if you try to make a request which has payload without mime-type, [ActionController TestCase](https://github.com/rails/rails/blob/master/actionpack/lib/action_controller/test_case.rb#L95) automatically adds `application/x-www-form-urlencoded` as content type to request header so making request without payload is not possible until changing this behaviour(AFAIK).
If I remove this line, lots of tests fail and I believe this will introduce another regression. Also the related [RFC](https://tools.ietf.org/html/rfc7231#section-3.1.1.5) says if you are sending a payload which **you know** the mime type of it, you **SHOULD** send the content type. Also the RFC says if the content-type header is missing receiver may assume it's content-type as `application/octet-stream`. In this case instead of changing the additions made in the commit I think we can treat the content which has no content-type information as default mime-type defined by community. Normally Rails returns nil if can't find content-type information [here](https://github.com/rails/rails/blob/master/actionpack/lib/action_dispatch/http/mime_negotiation.rb#L17).

@maclover7 what do you think?

**Comment by rafaelfranca** [comment]
I think our default was `application/x-www-form-urlencoded` too in Rails 4.2? Could someone investigate? We should match the same behavior

**Comment by mrjlynch** [comment]
I am experiencing this issue when using `capybara-webkit` and Rails `5.0.1`

I think I am getting this error after an ajax request when the response does not have a mime_type. 

Though these commits that fix this look like they were made before `5.0.1`?

**Comment by martijn** [comment]
@mrjlynch same here, that was fixed with https://github.com/rails/rails/pull/25950

**Comment by mrjlynch** [comment]
Thanks Martijn, rails 5.0.0.1 was released 10th August, #25950 was merged 26th July, thus shouldn't this not occur in 5.0.0.1?

**Comment by mrjlynch** [comment]
I have upgraded to rails `5.0.1.rc1` and this issue is resolved! 🎉
